File: javascript/0472-concatenated-words.js
/**
 * @param {string[]} words
 * @return {string[]}
 */
var findAllConcatenatedWordsInADict = function (words) {
    let wordSet = new Set(words);
    let res = [];

    for (let w of words) {
        if (dfs(w)) {
            res.push(w);
        }
    }

    return res;

    /**
     *
     * @param {string} word
     * @returns {boolean}
     */
    function dfs(word) {
        for (let i = 1; i < word.length; i++) {
            let prefix = word.slice(0, i);
            let suffix = word.slice(i, word.length);

            if (
                (wordSet.has(prefix) && wordSet.has(suffix)) ||
                (wordSet.has(prefix) && dfs(suffix))
            ) {
                return true;
            }
        }
        return false;
    }
};

File: go/0472-concatenated-words.go
func findAllConcatenatedWordsInADict(words []string) []string {
	wordSet := initWordSet(words)
	res := []string{}

	var dfs func(word string) bool

	dfs = func(word string) bool {
		for i := 1; i < len(word); i++ {
			prefix, suffix := word[:i], word[i:]
			if (wordSet[prefix] && wordSet[suffix]) || wordSet[prefix] && dfs(suffix) {
				return true
			}
		}

		return false
	}

	for _, w := range words {
		if dfs(w) {
			res = append(res, w)
		}
	}

	return res
}

func initWordSet(words []string) map[string]bool {
	wordSet := make(map[string]bool, len(words))

	for _, word := range words {
		wordSet[word] = true
	}

	return wordSet
}